### PR TITLE
LBAAS-2428 Support http idle timeout

### PR DESCRIPTION
**Pull request by StephenVarela**

### CODE REVIEW

File: cloud-controller-manager/do/loadbalancers.go [hunk 1]
```diff
@@ -1321,6 +1331,22 @@ func getEnableBackendKeepalive(service *v1.Service) (bool, error) {
...
+	httpIdleTimeoutSeconds, ok := service.Annotations[annDOHttpIdleTimeoutSeconds]
+
+	if !ok {
+		return nil, nil
+	}
+
+	httpIdleTimeout, err := strconv.ParseUint(httpIdleTimeoutSeconds, 10, 64)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse LB http_idle_timeout_seconds: %s", err)
```
Comment (anitgandhi) [inline cloud-controller-manager/do/loadbalancers.go:1344]:
```suggestion
		return nil, fmt.Errorf("failed to parse LB http_idle_timeout_seconds: %w", err)
```

File: cloud-controller-manager/do/loadbalancers.go [hunk 2]
```diff
@@ -1321,6 +1331,22 @@ func getEnableBackendKeepalive(service *v1.Service) (bool, error) {
 	return enableBackendKeepalive, nil
 }
 
+func getHttpIdleTimeoutSeconds(service *v1.Service) (*uint64, error) {
+	httpIdleTimeoutSeconds, ok := service.Annotations[annDOHttpIdleTimeoutSeconds]
+
```
Comment (anitgandhi) [inline cloud-controller-manager/do/loadbalancers.go:1336]:
```suggestion
```

File: cloud-controller-manager/do/loadbalancers.go [hunk 3]
```diff
@@ -1321,6 +1331,22 @@ func getEnableBackendKeepalive(service *v1.Service) (bool, error) {
...
+	if !ok {
+		return nil, nil
+	}
+
+	httpIdleTimeout, err := strconv.ParseUint(httpIdleTimeoutSeconds, 10, 64)
+
```
Comment (anitgandhi) [inline cloud-controller-manager/do/loadbalancers.go:1342]:
```suggestion
```

File: docs/controllers/services/annotations.md [hunk 4]
```diff
@@ -177,3 +177,7 @@ See also the [elaborate example](/docs/controllers/services/examples/README.md#c
 **Note**
 
 You have to supply the value as string (ex. `"true"`, not `true`), otherwise you might run into a [k8s bug that throws away all annotations on your `Service` resource](https://github.com/kubernetes/kubernetes/issues/59113).
+
+## service.beta.kubernetes.io/do-loadbalancer-http-idle-timeout-seconds
+
+Specifies the http idle timeout configuration in seconds. The ranges for this value are between 30 seconds and 10 minutes (600 seconds) The default value if not specified is 60 seconds
```
Comment (anitgandhi) [inline docs/controllers/services/annotations.md:183]:
```suggestion
Specifies the HTTP idle timeout configuration in seconds. If not specified, the default is 60 seconds.
```

**Review by anitgandhi — APPROVED** [review]

File: cloud-controller-manager/do/loadbalancers.go (hunk 1 above)
Comment (timoreimann) [reply cloud-controller-manager/do/loadbalancers.go:1344]: Do we actually intend to parse this error? If not or if it's unclear at this moment, then I think we should keep it at `%s` to not extend the API surface area unnecessarily per Go's recommendation: https://go.dev/blog/go1.13-errors#whether-to-wrap

Comment (anitgandhi) [reply cloud-controller-manager/do/loadbalancers.go:1344]: fair point 👍

File: cloud-controller-manager/do/loadbalancers.go [hunk 5]
```diff
@@ -1321,6 +1331,20 @@ func getEnableBackendKeepalive(service *v1.Service) (bool, error) {
...
 
+func getHttpIdleTimeoutSeconds(service *v1.Service) (*uint64, error) {
+	httpIdleTimeoutSeconds, ok := service.Annotations[annDOHttpIdleTimeoutSeconds]
+	if !ok {
+		return nil, nil
+	}
+
+	httpIdleTimeout, err := strconv.ParseUint(httpIdleTimeoutSeconds, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse LB http_idle_timeout_seconds: %s", err)
```
Comment (khayyamsaleem) [inline cloud-controller-manager/do/loadbalancers.go:1342]: would this error be more helpful in debugging if we specified the full annotation key associated with the unparseable value?

**Review by timoreimann — APPROVED** [review]

**Comment by timoreimann** [comment]
I see @khayyamsaleem raised a reasonable point. Holding off on merging for that discussion to resolve.

File: cloud-controller-manager/do/loadbalancers.go (hunk 5 above)
Comment (StephenVarela) [reply cloud-controller-manager/do/loadbalancers.go]: Added the full key like @anitgandhi suggested 🙏 

CC @khayyamsaleem  let me know if this looks okay

**Comment by timoreimann** [comment]
Took the liberty to merge as I think Khayyam's suggestion has been addressed.

**Pull request merged**